server/converters/cxgtool.py

```diff
@@ -0,0 +1,355 @@
+            )
+
+    else:
+        raise KeyError(f"Annotation {index_col_name}, specified in --{df_name}-name, does not exist.")
+
+    return (df, index_col_name)
+
+
+def save_dataframe(container, name, df, index_col_name, ctx):
+    A_name = f"{container}/{name}"
+    (df, index_col_name) = alias_index_col(df, name, index_col_name)
+    create_dataframe(A_name, df, ctx=ctx)
+    with tiledb.DenseArray(A_name, mode="w", ctx=ctx) as A:
+        value = {}
+        schema_hints = {}
+        for k, v in df.items():
+            dtype, hints = cxg_type(v)
+            value[k] = v.to_numpy(dtype=dtype)
+            if hints:
+                schema_hints.update({k: hints})
+
+        schema_hints.update({"index": index_col_name})
+        A[:] = value
+        A.meta["cxg_schema"] = json.dumps(schema_hints)
+
+    tiledb.consolidate(A_name, ctx=ctx)
+
+
+def create_emb(e_name, emb):
+    filters = tiledb.FilterList([tiledb.ZstdFilter(), ])
+    attrs = [tiledb.Attr(dtype=emb.dtype, filters=filters)]
+    dims = []
+    for d in range(emb.ndim):
+        shape = emb.shape
+        dims.append(tiledb.Dim("", domain=(0, shape[d] - 1), tile=min(shape[d], 1000), dtype=np.uint32))
+    domain = tiledb.Domain(*dims)
+    schema = tiledb.ArraySchema(
+        domain=domain, sparse=False, attrs=attrs, capacity=1_000_000, cell_order="row-major", tile_order="row-major"
+    )
+    tiledb.DenseArray.create(e_name, schema)
+
+
+def is_valid_embedding(adata, name, arr):
+    """ return True if this layout data is a valid array for front-end presentation:
+        * ndarray, with shape (n_obs, >= 2), dtype float/int/uint
+        * contains only finite values
+        * follows ScanPy embedding naming conventions
+    """
+    is_valid = type(name) == str and name.startswith("X_") and len(name) > 2
+    is_valid = is_valid and type(arr) == np.ndarray and arr.dtype.kind in "fiu"
+    is_valid = is_valid and arr.shape[0] == adata.n_obs and arr.shape[1] >= 2
+    is_valid = is_valid and np.all(np.isfinite(arr))
+    return is_valid
+
+
+def save_embeddings(container, adata, ctx):
+    for (name, value) in adata.obsm.items():
+        if is_valid_embedding(adata, name, value):
+            e_name = f"{container}/{name[2:]}"
+            create_emb(e_name, value)
+            with tiledb.DenseArray(e_name, mode="w", ctx=ctx) as A:
+                A[:] = value
+            log(1, f"\t\t...{name} embedding created")
+        tiledb.consolidate(e_name, ctx=ctx)
+
+
+def create_X(X_name, shape):
+    # Dense, always.  Future task: explore if sparse encoding is worth the trouble
+    # below a sparsity threshold.
+    filters = tiledb.FilterList([tiledb.ZstdFilter()])
+    attrs = [tiledb.Attr(dtype=np.float32, filters=filters)]
+    domain = tiledb.Domain(
+        tiledb.Dim(name="obs", domain=(0, shape[0] - 1), tile=min(shape[0], 50), dtype=np.uint32),
```

Comment:
There are some hard coded values here and elsewhere.  It might be worth either making them constants, and defining and commenting them at the top of the file.  Or making them command line options that have good defaults.

Reply:
for CLI params, lets leave these for a later pass, when we know if we really need them.   I'd like to keep parameterization to just what is going to be used.

I'll add some comments for magic numbers